Plugins/NativeCamera/Source/Android/NativeCameraImpl.cpp

```diff
@@ -51,35 +70,92 @@ namespace Babylon::Plugins
 
+        // Iterate over all of the cameras and find the one that has the best stream configuration
         for (int i = 0; i < cameraIds->numCameras; ++i)
         {
-            const char *id = cameraIds->cameraIds[i];
+            const char* id = cameraIds->cameraIds[i];
 
             API24::ACameraMetadata *metadataObj;
             GET_CAMERA_FUNCTION(ACameraManager_getCameraCharacteristics)(m_cameraManager, id, &metadataObj);
 
             API24::ACameraMetadata_const_entry lensInfo = {};
             GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_LENS_FACING, &lensInfo);
 
+            API24::ACameraMetadata_const_entry sensorOrientation = {};
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SENSOR_ORIENTATION, &sensorOrientation);
+
             auto facing = static_cast<API24::acamera_metadata_enum_android_lens_facing_t>(lensInfo.data.u8[0]);
+            if (facing != (frontCamera ? API24::ACAMERA_LENS_FACING_FRONT : API24::ACAMERA_LENS_FACING_BACK))
+            {
+                // Ignore cameras facing the wrong direction
+                continue;
+            }
+
+            // Get all available stream configurations supported by the camera
+            API24::ACameraMetadata_const_entry streamConfigurations;
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SCALER_AVAILABLE_STREAM_CONFIGURATIONS, &streamConfigurations);
+            // format of the data:
+            // 0: format
+            // 1: width
+            // 2: height
+            // 3: input?
+
+            for (uint32_t j = 0; j < streamConfigurations.count; j += 4) {
+                int32_t format = streamConfigurations.data.i32[j + 0];
+                int32_t width = streamConfigurations.data.i32[j + 1];
+                int32_t height = streamConfigurations.data.i32[j + 2];
+                int32_t input = streamConfigurations.data.i32[j + 3];
+
+                if (input || format != AIMAGE_FORMAT_YUV_420_888 || static_cast<uint32_t>(width) > maxWidth || static_cast<uint32_t>(height) > maxHeight)
+                {
+                    // Ignore the configuration if it is either an input type or not a preview format
+                    // Also ignore the configuration if either the width or height is beyond the max allowed
+                    continue;
+                }
 
-            // Found a corresponding facing camera?
-            if (facing == (frontCamera ? API24::ACAMERA_LENS_FACING_FRONT : API24::ACAMERA_LENS_FACING_BACK))
+                // Calculate pixel count and dimension differential and take the best qualifying one.
+                uint32_t pixelCount{static_cast<uint32_t>(width * height)};
+                uint32_t dimDiff{(maxWidth - width) + (maxHeight - height)};
+                if (bestCameraId == nullptr || pixelCount > bestPixelCount || (pixelCount == bestPixelCount && dimDiff < bestDimDiff))
```

Comment:
This is the same heuristic as iOS, right? Is it worth sharing code for this between the platforms? Like given a set of constraints and a set of supported resolutions, the behavior should be the same regardless of platform right? Right now the heuristic is very simple, so there isn't that much duplication, but if we were to support more of the constraints, then we would have more and more duplication. I guess we can refactor this later if we add support for more constraints.

Reply:
It was definitely something I was considering as I was implementing the Android specific side. I do think in the long run it would be ideal to have some common interface to talk about the available camera streams and their capabilities in order to compare them in a common implementation... but I wasn't planning to tackle that with this PR.